File: JavaScript/CalcStats/calcStats.js
var calcStats = {
  valueByCondition: function(nums, condition) {
    var min = nums[0];

    for (var i = 1; i < nums.length; i++) {
      if (condition(nums[i],min)) {
        min = nums[i];
      }
    }
    return min;
  },

  minValue: function(nums) {
    return this.valueByCondition(nums, (function(a, b) {return a < b}));
  },
  maxValue: function(nums) {
    return this.valueByCondition(nums, (function(a,b) {return a > b}));
  },
  numberOfElements: function(nums) {
    return nums.length;
  },
  average: function(nums) {
    sum = 0;
    for (var i = 0; i < nums.length; i++) {
      sum += nums[i];
    }
    return sum/this.numberOfElements(nums);
  }
};

// The following functions are for testing only

answerMin = function(nums) {
  return calcStats.minValue(nums);
}

answerMax = function(nums) {
  return calcStats.maxValue(nums);
}

answerNElements = function(nums) {
  return calcStats.numberOfElements(nums);
}

answerAverage = function(nums) {
  return calcStats.average(nums);
}

File: Haskell/CalcStats/CalcStats.hs
module CalcStats (minimumValue,
                  maximumValue,
                  numberOfElements,
                  average) where

minimumValue :: [Integer] -> Integer
maximumValue :: [Integer] -> Integer
numberOfElements :: [a] -> Integer

minimumValue (x:[]) = x
minimumValue (x:y:xs) = minimumValue ((min x y):xs)

maximumValue (x:[]) = x
maximumValue (x:y:xs) = maximumValue ((max x y):xs)

numberOfElements [] = 0
numberOfElements (x:[]) = 1
numberOfElements (x:xs) = 1 + (numberOfElements xs)

average xs =  (fromIntegral (sum xs)) / (fromIntegral (length xs))
